Add wigner plotting

thisac opened the pull request:
**Context:**
The Wigner plotting functionality currently on SF interactive should be moved to SF.

**Description of the Change:**
A new `plot` module is created with a `plot_wigner` function, that plots the Wigner function, along with a `generate_wigner_chart` function that returns the "plottable' Plot.ly dictionary.

**Benefits:**
Wigners can be plotted from SF.

**Possible Drawbacks:**
None

**Related GitHub Issues:**
None

josh146 commented on strawberryfields/plot.py:
```diff
@@ -0,0 +1,135 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+This module provides tools to visualize the state in various interactive
+ways using Plot.ly
```
Awesome 💪

josh146 commented on strawberryfields/plot.py:
```diff
@@ -0,0 +1,135 @@
+from copy import deepcopy
+
+import numpy as np
+import plotly.io as pio
```
We need to be careful with the import here, since plotly isn't a dependency of SF! If you have a look in `sf.apps.plot`, you can see how this is handled there (essentially, lots of try excepts)

josh146 commented on strawberryfields/plot.py:
```diff
@@ -0,0 +1,135 @@
+        wire (int): wire used to calculate the reduced Wigner function
+        dq (float): stepsize along the q and p axes
+    """
+    q = np.arange(-state.cutoff_dim, state.cutoff_dim, dq)
```
Will `cutoff_dim` be defined for Gaussian states?

We should probably allow the user to pass their _own_ value of the p and q axis to this function, for full control. E.g.,

```python
def plot_wigner(state, p_axis, q_axis, wires=0, renderer="browser")
```
(or something like that)

thisac replied:
> Will cutoff_dim be defined for Gaussian states?

No, this doesn't work with Gaussian states the way I wrote it. This was just a way SF interactive did it. I was planning to update this, but hadn't decided on exactly how to do it yet. Either just as you suggested or having the option to just call `plot_wigner(state)` for the Fock backend, which would use the cutoff (and raise an "must-define-p-and-q" error for the Gaussian), and otherwise having to write e.g. like your suggestion.

josh146 commented on strawberryfields/plot.py:
```diff
@@ -0,0 +1,135 @@
+    """Populates a chart dictionary with reduced Wigner function surface plot data.
+
+    Args:
+        qp (tuple[float]): minimum and maximum values of the q and p axes
```
I wonder how easy it would be to have this accept q and p axis list values as well 🤔

josh146 reviewed (commented):
Looking like it's in the right direction @thisac!

antalszava commented on strawberryfields/plot.py:
```diff
@@ -0,0 +1,152 @@
+        wire (int): wire used to calculate the reduced Wigner function
+        dq (float): stepsize along the p_axis and q_axis
+    """
+    xvec = np.arange(-q_axis, q_axis, dq)
+    pvec = np.arange(-p_axis, p_axis, dq)
```
Just a thought (correct me if it's not valid): could it make sense to pre-define the landscape of `x` and `p` values in advance? Or make their values be optional

josh146 replied:
It's a bit hard to pre-define the values without knowing (or at least, having a good guess) for the approximate domain of the state we are plotting 🤔 Maybe there is a rough heuristic between maximum photon number/cutoff dim and maximum quadrature value to plot?

@nquesada might know.

thisac replied:
Correct me if I'm wrong @josh146, but I think that the `[-cutoff, cutoff]` range was fairly arbitrarily chosen as a range that seemingly worked nicely. A better heuristic would be nice, but otherwise I feel like this could be left to the user to decide, and for the simulators we could just pass the cutoff values as ranges if we cannot think of a better way.

The other issue is that only the Fock backend would have a cutoff. So it would only work in that specific case then. In SF interactive we always pass a cutoff value, even when using another backend.

josh146 replied:
Yep, pretty much @thisac. I chose the `[-cutoff, cutoff]` domain very arbitrarily ~ 2 years ago. One downside is that for states with a large cutoff, the plotting domain is huge even if the state to be plotted is very close to the origin!

@antalszava I think this comment thread is now outdated -- the latest version of this PR (as far as I can tell) has the plotting function accept the domain directly as an argument.

nquesada replied:
I think it does not make much sense to allow for predefined ranges in phase space. It is certainly true that a state with `cutoff` Fock space support will be localized to a circle of radius of sqrt(cutoff) in phase space (Josh I think was being too generous two years ago). Yet, I think the burden of picking ranges should be passed to the user, deciding this in advance is just too tricky.

josh146 replied:
>Josh I think was being too generous two years ago

😆

antalszava commented on strawberryfields/plot.py:
```diff
@@ -0,0 +1,152 @@
+    chart["layout"]["paper_bgcolor"] = "white"
+    chart["layout"]["plot_bgcolor"] = "white"
+    chart["layout"]["scene"]["bgcolor"] = "white"
+    chart["layout"]["font"] = {"color": textcolor}
```
```suggestion
```

antalszava reviewed (commented):
Looking good to me so far! :) :100:

josh146 commented on strawberryfields/plot.py:
```diff
@@ -0,0 +1,152 @@
+    "plotly or visiting https://plot.ly/python/getting-started/#installation"
+)
+
+try:
+    import plotly.io as pio
+except ImportError as e:
+    raise(plotly_error) from e
```
It would be good to determine when this exception is raised. For example, we don't want this to be raised when SF is imported; we only want it to be raised if/when the user tries calling the `plot_wigner` function. Perhaps we could define a function:

```python
def _get_plotly():
    try:
        import plotly.io as pio
    except ImportError as e:
        raise(plotly_error) from e
    return pio
```

Then, within the plotting functions, we could do:

```python
def plot_wigner(state, p_axis, q_axis, renderer="browser", wire=0, dq=0.1):
    """Plot the Wigner function with Plot.ly"""
    pio = _get_plotly()
```

thisac replied:
I like this solution. Thanks @josh146!

josh146 commented on strawberryfields/plot.py:
```diff
@@ -0,0 +1,152 @@
+
+    Args:
+        state (:class:`.BaseState`): the state used for plotting
+        p_axis (int): Range of the discretized :math:`p` quadrature values.
+            E.g., ranging from ``-p_axis`` to ``p_axis``.
+        q_axis (int): Range of the discretized :math:`x` quadrature values.
+            E.g., ranging from ``-q_axis`` to ``q_axis``.
```
Shouldn't this be a float?

josh146 commented on strawberryfields/plot.py:
```diff
@@ -0,0 +1,152 @@
+        wire (int): wire used to calculate the reduced Wigner function
+        dq (float): stepsize along the p_axis and q_axis
+    """
+    xvec = np.arange(-q_axis, q_axis, dq)
+    pvec = np.arange(-p_axis, p_axis, dq)
```
There could be a case where the user wants to pass `xvec` or `pvec` directly. E.g., if the state to be plotted is far from the origin, you wouldn't want the axes to be symmetric around 0.

Perhaps it's better to just pass `xvec` and `pvec`, like you do with matplotlib, and like the `generate_wigner_chart` function.

nquesada reviewed (commented):
This is a great addition to SF. Wondering what happens if I ask for the Wigner function of a multimode state? Is this handled gracefully?

nquesada commented:
![image](https://user-images.githubusercontent.com/991946/101534741-9b9a6880-3965-11eb-8e10-3632d9038054.png)

This is how it look for a squeezed state. I'd suggest using `np.round` around the values in the `z` axis to minimize the unphysical riples.

josh146 commented:
> This is a great addition to SF. Wondering what happens if I ask for the Wigner function of a multimode state? Is this handled gracefully?

That's a very good point. I can think of two solutions for now:

1. Simply raise an exception if `state.num_modes > 1`.

2. Modify the plotting function to accept a compulsory `mode` argument. The reduced state of this mode is then plotted, e.g., `sf.plot_wigner(state, mode, xvec, pvec, renderer="browser")`.

I think option (2) is the better option, it avoids the user needing to compute the reduced state manually themselves.

Out of curiosity @nquesada, are there any visualization strategies for multi-mode Wigner functions?

thisac commented:
Thanks @nquesada and @josh146!

> This is how it look for a squeezed state. I'd suggest using np.round around the values in the z axis to minimize the unphysical riples.

Good point. Will fix this!

> 2. Modify the plotting function to accept a compulsory mode argument. The reduced state of this mode is then plotted, e.g., sf.plot_wigner(state, mode, xvec, pvec, renderer="browser").

This is actually almost the case already, although it's mode 0 is set to default and is called `wire` (`plot_wigner(state, xvec, pvec, renderer="browser", wire=0)`). Should this be compulsary instead do you think?

josh146 commented:
> This is actually almost the case already, although it's mode 0 is set to default and is called wire (plot_wigner(state, xvec, pvec, renderer="browser", wire=0)). Should this be compulsary instead do you think?

Yep, I think this should probably be compulsory. For comparison, the Wigner function method on the state has the signature `state.wigner(self, mode, xvec, pvec)`,

Also lets call it `mode`, to fit in with the rest of Strawberry Fields 😆

nquesada commented:
@josh146 : As far as I know, there are no easy ways to see object in four or more dimensions :(

thisac commented:
Changed `wire` to `mode` (too used to writing wires, so I didn't realize that this was in fact SF and not PL), and added a `contour=True` argument, so that contours can be turned off if wanted. 

Couldn't really figure out a nice way of removing the unphysical ripples though. `np.round` caused the rest of the image to become a bit blocky when over-rounding, and smaller values didn't really remove the ripples. Might need to think this over a bit more, but I think this might be fine for now.

nquesada commented:
@thisac : was the `mode` argument made compulsory for plotting?

thisac commented:
> was the mode argument made compulsory for plotting?

@nquesada Yes, it was!

nquesada commented on strawberryfields/plot.py:
```diff
@@ -0,0 +1,122 @@
+    return pio
+
+
+def plot_wigner(state, mode, xvec, pvec, renderer="browser", contours=True):
```
It is! Sorry I did not see it earlier.

nquesada commented:
Any idea how to get the coverage to not decrease?

josh146 commented on .github/CHANGELOG.md:
```diff
@@ -5,6 +5,25 @@
+  xvec = np.arange(-4, 4, 0.01)
+  pvec = np.arange(-4, 4, 0.01)
+
+  sf.plot_wigner(state, xvec, pvec, renderer="browser")
```
Needs the mode argument!
```suggestion
  sf.plot_wigner(state, 0, xvec, pvec, renderer="browser")
```

josh146 commented on strawberryfields/plot.py:
```diff
@@ -0,0 +1,122 @@
+        xvec (array): array of discretized :math:`x` quadrature values
+        pvec (array): array of discretized :math:`p` quadrature values
+        renderer (string): the renderer for plotting with Plot.ly
+        contours (bool): whether to show the contour lines in the plot
```
Awesome, I like this addition

josh146 commented on tests/frontend/test_sf_plot.py:
```diff
@@ -0,0 +1,54 @@
+        pvec = np.arange(-4, 4, 0.1)
+        with monkeypatch.context() as m:
+            m.setattr(pio, "show", lambda x: None)
+            plot_wigner(results.state, xvec, pvec, renderer=renderer, wire=wire)
```
This should be changed to `mode`, which makes me think this test is not being run...

Perhaps you need to add the `frontend` pytestmark to the file?

josh146 replied:
Also, it might be good to also add a small test checking that the generate chart data function produces the expected results!

josh146 commented:
> Any idea how to get the coverage to not decrease?

@nquesada in the apps layer, we simply marked the plotting functions with `# pragma: no cover` 🙈

antalszava commented on strawberryfields/plot.py:
```diff
@@ -0,0 +1,122 @@
+# limitations under the License.
+r"""
+This module provides tools to visualize the state in various interactive
+ways using Plot.ly
```
```suggestion
ways using Plot.ly.
```

antalszava commented on strawberryfields/plot.py:
```diff
@@ -0,0 +1,122 @@
+
+plotly_error = (
+    "Plot.ly required for using this function. It can be installed using pip install "
+    "plotly or visiting https://plot.ly/python/getting-started/#installation"
```
Wonder if it would be better to not include an explicit webpage here so that the URL doesn't have to be maintained (should it move or get removed).

thisac replied:
Good point, will remove. I just copied this from the apps `plot.py`.

antalszava commented on strawberryfields/plot.py:
```diff
@@ -0,0 +1,122 @@
+
+
+def _get_plotly():
+    """Import Plot.ly on demand to avoid errors being raised unnecessarily"""
```
```suggestion
    """Import Plot.ly on demand to avoid errors being raised unnecessarily."""
```

antalszava commented on .github/CHANGELOG.md:
````diff
@@ -5,6 +5,26 @@
+* Wigner functions can be plotted directly via Strawberry Fields using Plot.ly.
+  [(#495)](https://github.com/XanaduAI/strawberryfields/pull/495)
+
+  ```python
+  prog = sf.Program(1)
+  eng = sf.Engine('fock', backend_options={"cutoff_dim": 10})
+
+  with prog.context as q:
+    gamma = 2
+    Vgate(gamma) | q[0]
+
+  state = eng.run(prog).state
+
+  xvec = np.arange(-4, 4, 0.01)
+  pvec = np.arange(-4, 4, 0.01)
+  mode = 0
+
+  sf.plot_wigner(state, mode, xvec, pvec, renderer="browser")
````
This looks awesome rendered! :100:

antalszava reviewed (commented):
Looking good! :1st_place_medal: :blush:  Might check back for the remaining tests, but no blockers for me

josh146 commented on strawberryfields/plot.py:
```diff
@@ -0,0 +1,122 @@
+
+plotly_error = (
+    "Plot.ly required for using this function. It can be installed as follows:
+    "pip install plotly."
```
This was an old comment, might not be relevant any more!

antalszava commented:
[ch1777]
